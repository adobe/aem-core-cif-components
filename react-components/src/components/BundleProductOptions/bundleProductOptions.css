.customize {
    align-items: center;
    display: flex;
    grid-area: cart;
    justify-content: center;
}

.required_info {
    color: rgb(var(--venia-error));
}

.option_quantity_title {
    font-size: 1rem;
    font-weight: 600;
    line-height: 1.5;
    margin: 1.5rem 0 0.5rem;
}

.option_quantity_input {
    -moz-appearance: textfield;
    margin: 0;
    height: 34px;
    text-align: center;
    width: 54px;
    padding: 1px 9px;
}

.option_quantity_input::-webkit-outer-spin-button,
.option_quantity_input::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
}

.select_icons {
    --iconsBefore: 0;
    --iconsAfter: 1;
}

.select_options_root {
    width: 75%;
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-orient: vertical;
    -webkit-box-direction: normal;
    -ms-flex-direction: column;
    flex-direction: column;
}

.multiselect_options_root {
    background: white;
    border: 1px solid #707070;
    border-radius: 2px;
    color: #212121;
    font-size: 1rem;
    width: 75%;
    -webkit-appearance: none;
}

.customization__info {
    float: right;
}

.totalPrice {
    font-weight: 700;
}
