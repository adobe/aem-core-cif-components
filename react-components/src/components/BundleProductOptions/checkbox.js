/*******************************************************************************
 *
 *    Copyright 2020 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { array, shape, func, bool, number } from 'prop-types';
import Price from '../Price';

const Checkbox = props => {
    const { item, customization, options, handleSelectionChange } = props;

    const onChange = event => {
        const { checked, value } = event.target;
        let newCustomization;

        if (checked) {
            newCustomization = [...customization];
            const { id, price, quantity } = options.find(o => o.id == value);
            newCustomization.push({ id, price, quantity });
        } else {
            newCustomization = customization.filter(c => c.id != value);
        }

        handleSelectionChange(item.option_id, 1, newCustomization);
    };

    return (
        <>
            {options.map(o => (
                <div key={`option-${item.option_id}-${o.id}`} className="bundleProduct__option">
                    <label>
                        <input
                            type="checkbox"
                            checked={customization.findIndex(c => c.id === o.id) > -1}
                            onChange={onChange}
                            value={o.id}
                        />{' '}
                        {`${o.quantity} x ${o.label} +`}
                        <b>
                            <Price currencyCode={o.currency} value={o.price} />
                        </b>
                    </label>
                </div>
            ))}
        </>
    );
};

Checkbox.propTypes = {
    item: shape({
        required: bool.isRequired,
        option_id: number.isRequired
    }),
    customization: array.isRequired,
    options: array.isRequired,
    handleSelectionChange: func.isRequired
};

export default Checkbox;
