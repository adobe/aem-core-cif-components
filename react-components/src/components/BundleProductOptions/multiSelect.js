/*******************************************************************************
 *
 *    Copyright 2020 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { array, shape, func, bool, number } from 'prop-types';

const MultiSelect = props => {
    const { item, customization, options, handleSelectionChange } = props;

    const onChange = event => {
        const values = Array.from(event.target.options)
            .filter(e => e.selected)
            .map(e => e.value);
        const newCustomization = options
            .filter(o => values.includes(o.id.toString()))
            .map(o => {
                return { id: o.id, quantity: o.quantity, price: o.price };
            });
        handleSelectionChange(item.option_id, 1, newCustomization);
    };

    return (
        <select
            name={item.option_id}
            multiple="multiple"
            className="bundleProduct__option field__input"
            size="5"
            value={customization.map(c => c.id)}
            onChange={onChange}>
            {options.map(o => (
                <option key={`option-${item.option_id}-${o.id}`} value={o.id}>
                    {o.label}
                </option>
            ))}
        </select>
    );
};

MultiSelect.propTypes = {
    item: shape({
        required: bool.isRequired,
        option_id: number.isRequired
    }),
    customization: array.isRequired,
    options: array.isRequired,
    handleSelectionChange: func.isRequired
};

export default MultiSelect;
