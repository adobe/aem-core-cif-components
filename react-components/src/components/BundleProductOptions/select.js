/*******************************************************************************
 *
 *    Copyright 2020 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { array, shape, func, bool, number } from 'prop-types';
import { useTranslation } from 'react-i18next';

const Select = props => {
    const { item, customization, options, handleSelectionChange } = props;
    const { quantity } = item;
    const { can_change_quantity } =
        customization.length > 0 ? options.find(o => o.id === customization[0].id) : { can_change_quantity: false };
    const [t] = useTranslation('cart');

    const onChange = event => {
        const { value } = event.target;
        let newQuantity = 1;
        const newCustomization = options
            .filter(o => o.id == value)
            .map(o => {
                newQuantity = o.quantity;
                return { id: o.id, quantity: newQuantity, price: o.price };
            });

        handleSelectionChange(item.option_id, newQuantity, newCustomization);
    };

    const onQuantityChange = event => {
        try {
            const newQuantity = parseInt(event.target.value);
            if (newQuantity > 0) {
                handleSelectionChange(item.option_id, newQuantity, customization);
            }
        } catch {
            throw new Error('Invalid quantity entered');
        }
    };

    return (
        <>
            <div className="bundleOptionSelect__root">
                <span className="fieldIcons__root">
                    <span className="fieldIcons__input">
                        <select
                            aria-label={item.title}
                            className="select__input field__input bundleProduct__option"
                            name={item.id}
                            value={customization[0]?.id}
                            onChange={onChange}>
                            {!item.required && <option value="">None</option>}
                            {options.map(o => (
                                <option key={`option-${item.option_id}-${o.id}`} value={o.id}>
                                    {o.label}
                                </option>
                            ))}
                        </select>
                    </span>
                    <span className="fieldIcons__before"></span>
                    <span className="fieldIcons__after">
                        <span className="icon__root">
                            <svg
                                xmlns="http://www.w3.org/2000/svg"
                                width="18"
                                height="18"
                                viewBox="0 0 24 24"
                                fill="none"
                                stroke="currentColor"
                                strokeWidth="2"
                                strokeLinecap="round"
                                strokeLinejoin="round">
                                <polyline points="6 9 12 15 18 9"></polyline>
                            </svg>
                        </span>
                    </span>
                </span>
            </div>

            <h2 className="bundleProduct__quantityTitle">
                <span>{t('cart:quantity', 'Quantity')}</span>
            </h2>
            <input
                type="number"
                min="1"
                className="option__quantity"
                disabled={!can_change_quantity}
                value={quantity}
                onChange={onQuantityChange}
            />
        </>
    );
};

Select.propTypes = {
    item: shape({
        required: bool.isRequired,
        option_id: number.isRequired,
        quantity: number.isRequired
    }),
    customization: array.isRequired,
    options: array.isRequired,
    handleSelectionChange: func.isRequired
};

export default Select;
