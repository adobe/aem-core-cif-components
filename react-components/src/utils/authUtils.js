/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import { ApolloLink } from '@apollo/client';
import { checkCookie, cookieValue } from './cookieUtils';

const authLink = new ApolloLink((operation, forward) => {
    let token = checkCookie('cif.userToken') ? cookieValue('cif.userToken') : '';
    if (token.length > 0) {
        operation.setContext(({ headers }) => ({
            headers: {
                authorization: `Bearer ${token && token.length > 0 ? token : ''}`,
                ...headers
            }
        }));
    }

    return forward(operation);
});

export { authLink as graphqlAuthLink };
