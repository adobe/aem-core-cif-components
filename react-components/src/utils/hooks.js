/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import { useEffect, useCallback } from 'react';
import { useQuery } from '@apollo/client';
import { checkCookie, cookieValue } from './cookieUtils';
import { useApolloClient } from '@apollo/client';

import QUERY_COUNTRIES from '../queries/query_countries.graphql';

export const useEventListener = (target, type, listener, ...rest) => {
    useEffect(() => {
        target.addEventListener(type, listener, ...rest);

        // return a callback, which is called on unmount
        return () => {
            target.removeEventListener(type, listener, ...rest);
        };
    }, [listener, rest, target, type]);
};

export const useCookieValue = cookieName => {
    if (!cookieName || cookieName.length === 0) {
        return '';
    }
    let value = checkCookie(cookieName) ? cookieValue(cookieName) : '';
    const setCookieValue = (value, age) => {
        const cookieSettings = `path=/; domain=${window.location.hostname};Max-Age=${age !== undefined ? age : 3600}`;
        document.cookie = `${cookieName}=${value};${cookieSettings}`;
    };

    return [value, setCookieValue];
};

export const useCountries = () => {
    const { data, error } = useQuery(QUERY_COUNTRIES);
    if (error || !data || !data.countries) {
        return { error, countries: [] };
    }

    return { countries: data.countries };
};

/**
 * This hook is taken from the Peregrine library.
 * We don't use it because upgrading to the peregrine library that exports it would mean bringing in some dependencies we don't need (i.e. Redux)
 *
 * @param {DocumentNode} query - parsed GraphQL operation description
 *
 * @returns {Function} callback that runs the query and returns a Promise
 */
export const useAwaitQuery = query => {
    const apolloClient = useApolloClient();

    return useCallback(
        options => {
            return apolloClient.query({
                ...options,
                query
            });
        },
        [apolloClient, query]
    );
};

/**
 * This hook makes the query parameters of the URL available.
 */
export const useQueryParams = () => {
    // Better to use useLocation from react router here, but this doesn't work because of dependency mess up.
    return new URLSearchParams(window.location.search);
};
