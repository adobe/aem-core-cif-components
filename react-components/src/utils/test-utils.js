/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { render } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import { I18nextProvider } from 'react-i18next';

import i18n from '../../__mocks__/i18nForTests';
import ConfigContextProvider from '../context/ConfigContext';
import UserContextProvider from '../context/UserContext';

import mutationChangePassword from './mocks/mutationChangePassword';
import mutationCreateCustomer from './mocks/mutationCreateCustomer';
import mutationCreateDuplicateCustomer from './mocks/mutationCreateCustomerDuplicate';
import mutationDeleteCustomerAddress from './mocks/mutationDeleteCustomerAddress';
import mutationGenerateToken from './mocks/mutationGenerateToken';
import mutationGenerateTokenWrongPassword from './mocks/mutationGenerateTokenWrongPassword';
import mutationMergeCarts from './mocks/mutationMergeCarts';
import mutationPlaceOrder from './mocks/mutationPlaceOrder';
import mutationRevokeToken from './mocks/mutationRevokeToken';
import mutationShippingAddress from './mocks/mutationShippingAddress';
import queryCart from './mocks/queryCart';
import queryCountries from './mocks/queryCountries';
import queryCustomerCart from './mocks/queryCustomerCart';
import queryCustomerDetails from './mocks/queryCustomerDetails';
import queryCustomerInformation from './mocks/queryCustomerInformation';
import queryEmptyCart from './mocks/queryEmptyCart';
import queryNewCart from './mocks/queryNewCart';

const defaultMocks = [
    mutationChangePassword,
    mutationCreateCustomer,
    mutationCreateDuplicateCustomer,
    mutationDeleteCustomerAddress,
    mutationGenerateToken,
    mutationGenerateTokenWrongPassword,
    mutationMergeCarts,
    mutationPlaceOrder,
    mutationRevokeToken,
    mutationShippingAddress,
    queryCart,
    queryCountries,
    queryCustomerCart,
    queryCustomerDetails,
    queryCustomerInformation,
    queryEmptyCart,
    queryNewCart
];

const defaultConfig = {
    storeView: 'default',
    graphqlEndpoint: 'none'
};

// eslint-disable-next-line react/display-name
const allProviders = (config, userContext, mocks) => ({ children }) => {
    return (
        <MockedProvider mocks={mocks} addTypename={false}>
            <ConfigContextProvider config={config || defaultConfig}>
                <UserContextProvider initialState={userContext}>
                    <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
                </UserContextProvider>
            </ConfigContextProvider>
        </MockedProvider>
    );
};

/* Wrap all the React components tested with the library in a mocked Apollo provider */
const customRender = (ui, options = {}) => {
    const { config, userContext, mocks = defaultMocks, ...renderOptions } = options;
    return render(ui, { wrapper: allProviders(config, userContext, mocks), ...renderOptions });
};

export * from '@testing-library/react';
export { customRender as render };
