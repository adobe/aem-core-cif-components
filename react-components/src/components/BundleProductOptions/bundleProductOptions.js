/*******************************************************************************
 *
 *    Copyright 2020 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useConfigContext } from '../../context/ConfigContext';
import * as dataLayerUtils from '../../utils/dataLayerUtils';
import Checkbox from './checkbox';
import Radio from './radio';
import Select from './select';
import MultiSelect from './multiSelect';

import { useAwaitQuery } from '../../utils/hooks';

import Button from '../Button';

import BUNDLE_PRODUCT_QUERY from '../../queries/query_bundle_product.graphql';
import Price from '../Price';

const BundleProductOptions = () => {
    const [t] = useTranslation(['product', 'cart']);
    const bundleProductQuery = useAwaitQuery(BUNDLE_PRODUCT_QUERY);
    const {
        mountingPoints: { bundleProductOptionsContainer }
    } = useConfigContext();
    const sku = document.querySelector(bundleProductOptionsContainer)?.dataset?.sku;
    const productId = document.querySelector('[data-cmp-is=product]')?.id;
    const [bundleState, setBundleState] = useState(null);

    const fetchBundleDetails = async sku => {
        const { data, error } = await bundleProductQuery({ variables: { sku }, fetchPolicy: 'network-only' });

        if (error) {
            throw new Error(error);
        }
        let currencyCode;
        const bundleOptions = data.products.items[0];
        const selections = bundleOptions.items.map(item => {
            return {
                option_id: item.option_id,
                required: item.required,
                title: item.title,
                type: item.type,
                quantity: ['checkbox', 'multi'].includes(item.type) ? 1 : item.options.find(o => o.is_default).quantity,
                options: item.options
                    .slice() // return a shallow copy of the array. the original is frozen and cannot be sorted as Array.prototype.sort() sorts the elements of the array in place
                    .sort((a, b) => a.position - b.position)
                    .map(option => {
                        const {
                            id,
                            can_change_quantity,
                            quantity,
                            label,
                            product: {
                                price_range: {
                                    maximum_price: {
                                        final_price: { currency, value }
                                    }
                                }
                            }
                        } = option;

                        return {
                            id,
                            can_change_quantity,
                            quantity,
                            label,
                            currency,
                            price: value
                        };
                    }),
                customization: item.options
                    .filter(o => o.is_default)
                    .map(o => {
                        const {
                            id,
                            quantity,
                            product: {
                                price_range: {
                                    maximum_price: {
                                        final_price: { value, currency }
                                    }
                                }
                            }
                        } = o;

                        currencyCode = currency;

                        return {
                            id,
                            quantity,
                            price: value
                        };
                    })
            };
        });
        setBundleState({ selections, currencyCode, quantity: 1 });
    };

    const handleSelectionChange = (option_id, quantity, customization) => {
        const { selections } = bundleState;
        const selectionIndex = selections.findIndex(s => s.option_id === option_id);
        const selection = selections[selectionIndex];
        selections.splice(selectionIndex, 1, { ...selection, quantity, customization });

        setBundleState({ ...bundleState, selections: [...selections] });
    };

    const changeBundleQuantity = e => {
        const { value } = e.target;
        setBundleState({ ...bundleState, quantity: value });
    };

    const renderItemOptions = item => {
        const { option_id, required, quantity, options, type, customization } = item;
        const otherProps = { options, customization, handleSelectionChange };

        switch (type) {
            case 'checkbox': {
                return <Checkbox item={{ option_id, required }} {...otherProps} />;
            }
            case 'radio': {
                return <Radio item={{ option_id, required, quantity }} {...otherProps} />;
            }
            case 'select': {
                return <Select item={{ option_id, required, quantity }} {...otherProps} />;
            }
            case 'multi': {
                return <MultiSelect item={{ option_id, required }} {...otherProps} />;
            }
        }
    };

    const canAddToCart = () => {
        const { selections } = bundleState;
        return selections.reduce((acc, selection) => {
            return acc && (selection.required ? selection.customization.length > 0 : true);
        }, true);
    };

    const addToCart = () => {
        const { selections, quantity } = bundleState;
        const productData = {
            sku,
            virtual: false,
            bundle: true,
            quantity: quantity,
            options: selections.map(s => {
                return {
                    id: s.option_id,
                    quantity: s.quantity,
                    value: s.customization.map(c => c.id.toString())
                };
            })
        };
        const customEvent = new CustomEvent('aem.cif.add-to-cart', {
            detail: [productData]
        });
        document.dispatchEvent(customEvent);
        // https://github.com/adobe/xdm/blob/master/docs/reference/datatypes/productlistitem.schema.md
        dataLayerUtils.pushEvent('cif:addToCart', {
            '@id': productId,
            'xdm:SKU': productData.sku,
            'xdm:quantity': productData.quantity,
            bundle: true
        });
    };

    const getTotalPrice = () => {
        const { selections, currencyCode } = bundleState;
        const price = selections.reduce((acc, selection) => {
            return (
                acc +
                selection.quantity *
                    selection.customization.reduce((a, c) => {
                        return a + c.price * (['checkbox', 'multi'].includes(selection.type) ? c.quantity : 1);
                    }, 0)
            );
        }, 0);

        return <Price currencyCode={currencyCode} value={price} className="bundlePrice" />;
    };

    if (!sku) return <></>;

    if (bundleState === null) {
        return (
            <section className="productFullDetail__section productFullDetail__customizeBundle">
                <Button priority="high" onClick={() => fetchBundleDetails(sku)}>
                    <span>{t('product:customize-bundle', 'Customize')}</span>
                </Button>
            </section>
        );
    }

    return (
        <>
            {bundleState.selections.map(e => (
                <section
                    key={`item-${e.option_id}`}
                    className="productFullDetail__section productFullDetail__bundleProduct">
                    <h3 className="option__title">
                        <span>{e.title}</span> {e.required && <span className="required"> *</span>}
                    </h3>
                    <div>{renderItemOptions(e)}</div>
                </section>
            ))}
            <section className="productFullDetail__section productFullDetail__bundleProduct">
                <h3>
                    <span className="required">* {t('product:required-fields', 'Required fields')}</span>
                    <span className="priceInfo">
                        {t('product:customization-price', 'Your customization')}: {getTotalPrice()}
                    </span>
                </h3>
            </section>
            <section className="productFullDetail__quantity productFullDetail__section">
                <h2 className="productFullDetail__quantityTitle option__title">
                    <span>{t('cart:quantity', 'Quantity')}</span>
                </h2>
                <div className="quantity__root">
                    <span className="fieldIcons__root" style={{ '--iconsBefore': 0, '--iconsAfter': 1 }}>
                        <span className="fieldIcons__input">
                            <select
                                aria-label="product's quantity"
                                className="select__input field__input"
                                name="quantity"
                                value={bundleState.quantity}
                                onChange={changeBundleQuantity}>
                                <option value="1">1</option>
                                <option value="2">2</option>
                                <option value="3">3</option>
                                <option value="4">4</option>
                            </select>
                        </span>
                        <span className="fieldIcons__before"></span>
                        <span className="fieldIcons__after">
                            <span className="icon__root">
                                <svg
                                    xmlns="http://www.w3.org/2000/svg"
                                    width="18"
                                    height="18"
                                    viewBox="0 0 24 24"
                                    fill="none"
                                    stroke="currentColor"
                                    strokeWidth="2"
                                    strokeLinecap="round"
                                    strokeLinejoin="round">
                                    <polyline points="6 9 12 15 18 9"></polyline>
                                </svg>
                            </span>
                        </span>
                    </span>
                    <p className="message-root"></p>
                </div>
            </section>
            <section className="productFullDetail__cartActions productFullDetail__section">
                <button
                    className="button__root_highPriority button__root clickable__root button__filled"
                    type="button"
                    disabled={!canAddToCart()}
                    onClick={addToCart}>
                    <span className="button__content">
                        <span>{t('product:add-item', 'Add to Cart')}</span>
                    </span>
                </button>
            </section>
        </>
    );
};

export default BundleProductOptions;
