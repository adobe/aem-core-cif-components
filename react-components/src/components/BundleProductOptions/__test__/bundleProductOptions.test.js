/*******************************************************************************
 *
 *    Copyright 2020 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { fireEvent, waitForDomChange } from '@testing-library/react';
import { render } from 'test-utils';
import * as hooks from '../../../utils/hooks';
import BundleProductOptions from '../bundleProductOptions';
import mockResponse from './graphQlMockReponse';

const config = {
    storeView: 'default',
    graphqlEndpoint: 'endpoint',
    mountingPoints: {
        bundleProductOptionsContainer: '#bundle-product-options'
    }
};

describe('<BundleProductOptions>', () => {
    it('renders the component with no sku', () => {
        const { asFragment } = render(<BundleProductOptions />, { config: config });
        expect(asFragment()).toMatchInlineSnapshot(`<DocumentFragment />`);
    });

    it('renders the component with sku', () => {
        const bundleProductOptionsContainer = document.createElement('div');
        bundleProductOptionsContainer.dataset.sku = 'VA24';
        bundleProductOptionsContainer.id = 'bundle-product-options';
        const { asFragment } = render(<BundleProductOptions />, {
            config: config,
            container: document.body.appendChild(bundleProductOptionsContainer)
        });

        expect(asFragment()).toMatchSnapshot();
    });

    it('renders the component with full options', async () => {
        // mock useState to return the state for a full rendering
        jest.spyOn(hooks, 'useAwaitQuery').mockImplementation(() => {
            return jest.fn().mockImplementation(async () => {
                return {
                    data: mockResponse,
                    error: null
                };
            });
        });

        const dispatchEventSpy = jest.spyOn(document, 'dispatchEvent').mockImplementation(event => {
            return event.detail;
        });

        const bundleProductOptionsContainer = document.createElement('div');
        bundleProductOptionsContainer.dataset.sku = 'VA24';
        bundleProductOptionsContainer.id = 'bundle-product-options';

        const { asFragment, container, getByRole } = render(<BundleProductOptions />, {
            config: config,
            container: document.body.appendChild(bundleProductOptionsContainer)
        });

        fireEvent.click(getByRole('button', { name: 'Customize' }));

        await waitForDomChange({ container });
        expect(asFragment()).toMatchSnapshot();

        fireEvent.click(getByRole('button', { name: 'Add to Cart' }));

        // Remove a required item to disable Add to Cart
        fireEvent.click(getByRole('checkbox', { name: '1 x Carmina Necklace + $78.00' }));
        fireEvent.click(getByRole('button', { name: 'Add to Cart' }));

        // Add to cart should be called just once since the second click was on a disabled button
        expect(dispatchEventSpy).toHaveBeenCalledTimes(1);

        // The mock dispatchEvent function returns the CustomEvent detail
        expect(dispatchEventSpy).toHaveReturnedWith([
            {
                bundle: true,
                options: [
                    { id: 1, quantity: 1, value: ['1'] },
                    { id: 2, quantity: 1, value: ['3'] },
                    { id: 3, quantity: 1, value: ['5'] },
                    { id: 4, quantity: 1, value: ['7', '8'] }
                ],
                quantity: 1,
                sku: 'VA24',
                virtual: false
            }
        ]);
    });
});
