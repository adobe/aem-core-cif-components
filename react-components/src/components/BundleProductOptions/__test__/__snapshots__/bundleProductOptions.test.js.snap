// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BundleProductOptions> renders the component with full options 1`] = `
<DocumentFragment>
  <section
    class="productFullDetail__section productFullDetail__bundleProduct"
  >
    <h3
      class="option__title"
    >
      <span>
        Necklace
      </span>
       
      <span
        class="required"
      >
         *
      </span>
    </h3>
    <div>
      <div
        class="bundleProduct__option"
      >
        <label>
          <input
            checked=""
            type="checkbox"
            value="1"
          />
           1 x Carmina Necklace +
          <b>
            <span
              class=""
            >
              $78.00
            </span>
          </b>
        </label>
      </div>
      <div
        class="bundleProduct__option"
      >
        <label>
          <input
            type="checkbox"
            value="2"
          />
           1 x Augusta Necklace +
          <b>
            <span
              class=""
            >
              $98.00
            </span>
          </b>
        </label>
      </div>
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__bundleProduct"
  >
    <h3
      class="option__title"
    >
      <span>
        Cirque Earrings
      </span>
       
      <span
        class="required"
      >
         *
      </span>
    </h3>
    <div>
      <div
        class="bundleOptionSelect__root"
      >
        <span
          class="fieldIcons__root"
        >
          <span
            class="fieldIcons__input"
          >
            <select
              class="select__input field__input bundleProduct__option"
            >
              <option
                value="3"
              >
                Gold Cirque Earrings
              </option>
              <option
                value="4"
              >
                Silver Cirque Earrings
              </option>
            </select>
          </span>
          <span
            class="fieldIcons__before"
          />
          <span
            class="fieldIcons__after"
          >
            <span
              class="icon__root"
            >
              <svg
                fill="none"
                height="18"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polyline
                  points="6 9 12 15 18 9"
                />
              </svg>
            </span>
          </span>
        </span>
      </div>
      <h2
        class="bundleProduct__quantityTitle"
      >
        <span>
          Quantity
        </span>
      </h2>
      <input
        class="option__quantity"
        min="1"
        type="number"
        value="1"
      />
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__bundleProduct"
  >
    <h3
      class="option__title"
    >
      <span>
        Sol Earrings
      </span>
       
      <span
        class="required"
      >
         *
      </span>
    </h3>
    <div>
      <div
        class="bundleProduct__options"
      >
        <label>
          <input
            checked=""
            name="3"
            type="radio"
            value="5"
          />
           Gold Sol Earrings +
          <b>
            <span
              class=""
            >
              $48.00
            </span>
          </b>
        </label>
      </div>
      <div
        class="bundleProduct__options"
      >
        <label>
          <input
            name="3"
            type="radio"
            value="6"
          />
           Silver Sol Earrings +
          <b>
            <span
              class=""
            >
              $48.00
            </span>
          </b>
        </label>
      </div>
      <h2
        class="option__title productFullDetail__quantityTitle"
      >
        <span>
          Quantity
        </span>
      </h2>
      <input
        class="option__quantity"
        disabled=""
        min="1"
        type="number"
        value="1"
      />
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__bundleProduct"
  >
    <h3
      class="option__title"
    >
      <span>
        Bangles
      </span>
       
      <span
        class="required"
      >
         *
      </span>
    </h3>
    <div>
      <select
        class="bundleProduct__option field__input"
        multiple=""
        name="4"
        size="5"
      >
        <option
          value="7"
        >
          Gold Omni Bangle Set
        </option>
        <option
          value="8"
        >
          Semper Bangle Set
        </option>
        <option
          value="9"
        >
          Silver Amor Bangle Set
        </option>
      </select>
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__bundleProduct"
  >
    <h3>
      <span
        class="required"
      >
        * Required fields
      </span>
      <span
        class="priceInfo"
      >
        Your customization: 
        <span
          class="bundlePrice"
        >
          $380.00
        </span>
      </span>
    </h3>
  </section>
  <section
    class="productFullDetail__quantity productFullDetail__section"
  >
    <h2
      class="productFullDetail__quantityTitle option__title"
    >
      <span>
        Quantity
      </span>
    </h2>
    <div
      class="quantity__root"
    >
      <span
        class="fieldIcons__root"
      >
        <span
          class="fieldIcons__input"
        >
          <select
            aria-label="product's quantity"
            class="select__input field__input"
            name="quantity"
          >
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
            <option
              value="4"
            >
              4
            </option>
          </select>
        </span>
        <span
          class="fieldIcons__before"
        />
        <span
          class="fieldIcons__after"
        >
          <span
            class="icon__root"
          >
            <svg
              fill="none"
              height="18"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="6 9 12 15 18 9"
              />
            </svg>
          </span>
        </span>
      </span>
      <p
        class="message-root"
      />
    </div>
  </section>
  <section
    class="productFullDetail__cartActions productFullDetail__section"
  >
    <button
      class="button__root_highPriority button__root clickable__root button__filled"
      type="button"
    >
      <span
        class="button__content"
      >
        <span>
          Add to Cart
        </span>
      </span>
    </button>
  </section>
</DocumentFragment>
`;

exports[`<BundleProductOptions> renders the component with sku 1`] = `
<DocumentFragment>
  <section
    class="productFullDetail__section productFullDetail__customizeBundle"
  >
    <button
      class="root_highPriority"
      type="button"
    >
      <span
        class="content"
      >
        <span>
          Customize
        </span>
      </span>
    </button>
  </section>
</DocumentFragment>
`;
